src/main/java/org/spongepowered/api/util/blockray/RayTrace.java

```diff
@@ -0,0 +1,339 @@
+import org.spongepowered.api.util.AABB;
+import org.spongepowered.api.world.Locatable;
+import org.spongepowered.api.world.LocatableBlock;
+import org.spongepowered.api.world.ServerLocation;
+import org.spongepowered.api.world.server.ServerWorld;
+import org.spongepowered.math.vector.Vector3d;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+
+/**
+ * Builds a set of parameters to define a ray trace that traces from a source
+ * position to a target position (or in a specified direction), subject to the
+ * supplied constraints.
+ *
+ * <p>A ray trace will check supplied predicates in the following order in the
+ * case where multiple predicates should be checked:</p>
+ *
+ * <ol>
+ *    <li>{@link #continueWhileLocation(Predicate)} -- called when the boundary
+ *    between two blocks is crossed, checking if the {@link ServerLocation} the
+ *    ray is crossing in to is valid, regardless of what the ray has otherwise
+ *    hit.</li>
+ *    <li>{@link #select(Predicate)} -- called when the {@link AABB} of
+ *    {@code T} is crossed by the ray trace, for {@link LocatableBlock}s this
+ *    is called directly after {@link #continueWhileLocation(Predicate)}. This
+ *    is the only predicate that can return a result, <strong>any</strong> .</li>
+ *    <li>{@link #continueWhileBlock(Predicate)} -- called if no selection was
+ *    returned (if appropriate), checks the {@link LocatableBlock} the ray is
+ *    entering. All predicates should return {@code true} if the ray should
+ *    continue to trace through the block.</li>
+ *    <li>{@link #continueWhileEntity(Predicate)} -- called if no selection was
+ *    returned (if appropriate), checks the {@link Entity} the ray has struck.
+ *    All predicates should return {@code true} if the ray should continue to
+ *    trace through the entity.</li>
+ * </ol>
+ *
+ * @param <T> The type of {@link Locatable} that this ray trace will attempt to
+ *            select
+ */
+public interface RayTrace<T extends Locatable> {
+
+    /**
+     * Creates a {@link RayTrace} that will attempt to select a
+     * {@link LocatableBlock}.
+     *
+     * @return The ray trace builder.
+     */
+    static RayTrace<LocatableBlock> block() {
+        return Sponge.getRegistry().getFactoryRegistry().provideFactory(Factory.class).blockRayTrace();
+    }
+
+    /**
+     * Creates a {@link RayTrace} that will attempt to select an {@link Entity}.
+     *
+     * @return The ray trace builder.
+     */
+    static RayTrace<Entity> entity() {
+        return Sponge.getRegistry().getFactoryRegistry().provideFactory(Factory.class).entityRayTrace();
+    }
+
+    /**
+     * Returns a {@link Predicate} for use in
+     * {@link #continueWhileLocation(Predicate)} that only allows a trace to
+     * continue if the {@link LocatableBlock} at the given position is air.
+     *
+     * @return The predicate
+     */
+    static Predicate<ServerLocation> onlyAir() {
```

Comment:
The switch of returns are a hair confusing here. What is the reasoning for that?

Reply:
Actually, this is a good catch in itself as this really should be `LocatableBlock` for the purpose that I was intending to use it for.

The intention behind this method and the next one (which should be `notAir`) is to create a ray that selects the first non-air block. For that to work given the javadocs, this would need to be set in `continueWhileBlock`, while `select` would then take a `notAir()` `Predicate<LocatableBlock>` - that is, continue the ray until we hit a non air block, selecting the non air block (`select` occurs before `continueWhileBlock`). There may also be another helper method for stopping a ray if an `Entity` is hit, though that is as simple as calling `continueWhileEntity(r -> false);`.

My development of this went through a few iterations and this looks like it was left behind. These will get adjusted.